lib/streamlit/elements/metric.py

```diff
@@ -13,24 +13,40 @@
 # limitations under the License.
 
 from textwrap import dedent
-from typing import Optional, cast
+from typing import cast, TYPE_CHECKING, Union
+from typing_extensions import TypeAlias, Literal
 
 import attr
-import streamlit
 from streamlit.errors import StreamlitAPIException
 from streamlit.proto.Metric_pb2 import Metric as MetricProto
 
 from .utils import clean_text
 
+if TYPE_CHECKING:
+    import numpy as np
 
-@attr.s(auto_attribs=True, slots=True)
+    from streamlit.delta_generator import DeltaGenerator
+
+
+Value: TypeAlias = Union["np.integer", "np.floating", float, str, None]
+Delta: TypeAlias = Union[float, str, None]
+DeltaColor: TypeAlias = Literal["normal", "inverse", "off"]
+
+
+@attr.s(auto_attribs=True, slots=True, frozen=True)
 class MetricColorAndDirection:
-    color: Optional[int]
-    direction: Optional[int]
+    color: MetricProto.MetricColor.ValueType
+    direction: MetricProto.MetricDirection.ValueType
 
 
 class MetricMixin:
-    def metric(self, label, value, delta=None, delta_color="normal"):
+    def metric(
+        self,
+        label: str,
+        value: Value,
+        delta: Delta = None,
+        delta_color: DeltaColor = "normal",
+    ) -> str:
```

Comment:
Was it intentional that this should return a string and not a `DeltaGenerator` like other non-io widgets?

Reply:
Hm, I'd guess that was just an error that somehow managed to slip by, but I'm not entirely sure there isn't some weird reason that this was intentional.

CC @willhuang1997

Reply:
Could you explain why?

Returning a serialized-to-string delta generator here results in something like this being returned:

```python
DeltaGenerator(_root_container=0, _provided_cursor=LockedCursor(_root_container=0, _index=0, _parent_path=(), _props={'delta_type': 'metric', 'last_index': None}), _parent=DeltaGenerator(_root_container=0, _provided_cursor=None, _parent=None, _block_type=None, _form_data=None), _block_type=None, _form_data=None)
```

which I can't see as anything that could possibly ever be considered helpful for the user.